optaplanner-examples/src/test/java/org/optaplanner/examples/nqueens/solver/score/NQueensConstraintProviderTest.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.examples.nqueens.solver.score;
+
+import org.junit.Test;
+import org.optaplanner.examples.nqueens.domain.Column;
+import org.optaplanner.examples.nqueens.domain.NQueens;
+import org.optaplanner.examples.nqueens.domain.Queen;
+import org.optaplanner.examples.nqueens.domain.Row;
+import org.optaplanner.test.impl.score.stream.ConstraintVerifier;
+import org.optaplanner.test.impl.score.stream.SingleConstraintVerifier;
+
+public class NQueensConstraintProviderTest {
+
+    private final NQueensConstraintProvider constraintProvider = new NQueensConstraintProvider();
+    private final ConstraintVerifier<NQueens> constraintVerifier =
+            ConstraintVerifier.createFor(NQueens.class, Queen.class);
+
+    @Test
+    public void testSingleConstraint() {
+        Column column1 = new Column();
+        column1.setId(0L);
+        column1.setIndex(0);
+        Row row = new Row();
+        row.setId(0L);
+        row.setIndex(0);
+        Queen queen1 = new Queen();
+        queen1.setId(0L);
+        queen1.setRow(row);
+        queen1.setColumn(column1);
+        // One queen
+        SingleConstraintVerifier<NQueens> horizontalConflictConstraintVerifier =
+                constraintVerifier.forConstraint(constraintProvider::horizontalConflict);
+        horizontalConflictConstraintVerifier.givenFacts(queen1, row, column1)
+                .expectImpact("No horizontal conflicts with just one queen.", 0);
+        // Two queens
+        Column column2 = new Column();
+        column2.setId(1L);
+        column2.setIndex(1);
+        Queen queen2 = new Queen();
+        queen2.setId(1L);
+        queen2.setRow(row);
+        queen2.setColumn(column2);
+        horizontalConflictConstraintVerifier.givenFacts(queen1, queen2, row, column1, column2)
+                .expectImpact("One pair of queens on the same row.", 1);
+        // Three queens
+        Column column3 = new Column();
+        column2.setId(2L);
+        column2.setIndex(2);
+        Queen queen3 = new Queen();
+        queen3.setId(2L);
+        queen3.setRow(row);
+        queen3.setColumn(column3);
+        horizontalConflictConstraintVerifier.givenFacts(queen1, queen2, queen3, row, column1, column2, column3)
+                .expectImpact("Three pairs of queens on the same row.", 3);
```

Comment:
If `3` is the total match weight, I think it should be `SimpleScore.of(3)`. I think it will make the tests easier to understand although it requires more keystrokes than just providing the number.

Reply:
I don't know what you mean. You've all made it very clear to me during the last meeting that we should not match scores, as constraint weights are externally configurable.

Reply:
I'm not talking about the total score. I'm only talking about the match weight.

I went through the code again and I think I've probably confused constraint weight and match weight (in the API, not conceptually). So if I have:

```java

    Constraint horizontalConflict(ConstraintFactory factory) {
        return factory
                .fromUniquePair(Queen.class, equal(Queen::getRowIndex))
                .penalize("Horizontal conflict", SimpleScore.ONE);
    }
```
`SimpleScore.ONE` is constraint weight, which only determines score level and the base that can be weighted. In this case match weight is not provided, so I guess it defaults to 1.

Whereas:
```java
    Constraint computerCost(ConstraintFactory constraintFactory) {
        return constraintFactory.from(CloudComputer.class)
                .ifExists(CloudProcess.class, equal(Function.identity(), CloudProcess::getComputer))
                .penalize("computerCost",
                        HardSoftScore.ONE_SOFT,
                        CloudComputer::getCost);
    }
```
in this case constraint weight is `HardSoftScore.ONE_SOFT` and it's given an additional (dynamic) weight of `CloudComputer::getCost`.

If this is true then the question in the beginning of this thread doesn't make sense. Result of weight calculation is a number (`int`, `long`, `Big*`) so it makes sense as it is (`expectImpact(String msg, Number totalWeight)`).

Reply:
Yes, exactly. The score (_HardSoftScore.ONE_SOFT_) is called "constraint weight", and can be configured externally using `ConstraintConfiguration`. The other value, often calculated (`CloudComputer::getCost`), is called "match weight", it is a numeric type and is defined within the constraint itself.